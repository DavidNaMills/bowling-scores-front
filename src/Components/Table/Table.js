import React from 'react';
import classes from './Table.module.scss';

const Table = ({ caption = '', tableClass = classes.table, data, showRowNum = false, selectRow = () => { }, ...rest }) => (
    <div>
        {caption && <p className={classes.caption}>{caption}</p>}
        <table className={tableClass}>
            <thead>
                {showRowNum && <th></th>}
                {data.headers.map((x, i) =>
                    <th key={i}>
                        {x}
                    </th>
                    , 0)}
            </thead>
            <tbody>
                {
                    data.rows.map((x, i) =>
                        <tr
                            key={i}
                            className={x.style && typeof (x.style.trow) === 'string' ? x.style.trow : ''}
                            style={x.style && typeof (x.style.trow) === 'object' ? x.style.trow : {}}
                            onClick={() => selectRow(x.id ? x.id : i + 1)}
                        >
                            {showRowNum && <td key={`${i}0`}>{i + 1}</td>}
                            {
                                x.values.map((y, o) =>
                                    <td key={`${i}-${o}`}>
                                        {y}
                                    </td>
                                    , 1)
                            }
                        </tr>
                    )
                }
            </tbody>
        </table>
    </div>
)

export default Table;

// const tHeaders = ['name', 'Average', 'Pinfall', 'Rank'];
// const tData = [
//     {
//         id: '',
//         values: ['david', 135, 2445, '+ 2'],
//         style: { tr: classes.testRowClass }  //optional
//     },
//     {
//         id: '5vd5f3d5cvx3cv5xz53xcz',
//         values: ['alan', 189, 2222, '-2'],
//         style: { trow: { color: 'red' } }  //optional
//     }
// ];