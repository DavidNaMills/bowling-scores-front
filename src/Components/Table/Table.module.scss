.table{
    width: 100%;
    border: solid black 1px;
    border-collapse: collapse;
    color: black;
    

    thead {
        
        background: rgb(255,255,255);
        background: -moz-linear-gradient(180deg, rgba(255,255,255,1) 37%, rgba(201,201,201,1) 100%);
        background: -webkit-linear-gradient(180deg, rgba(255,255,255,1) 37%, rgba(201,201,201,1) 100%);
        background: linear-gradient(180deg, rgba(255,255,255,1) 37%, rgba(201,201,201,1) 100%);
        filter: progid:DXImageTransform.Microsoft.gradient(startColorstr="#ffffff",endColorstr="#c9c9c9",GradientType=1);
    }

    th{
        font-size: 16px;
        border: solid rgb(111, 111, 111) 1px;
        color: black;
        padding: 5px;
    }
    
    td{
        font-size: 14px;
        border: solid rgb(111, 111, 111) 1px;
        padding: 5px;
        text-align: center;
    }  
}

.caption{
    font-size: 14px;
    padding:0;
    margin-top:15px;
    margin-bottom:5px;
}