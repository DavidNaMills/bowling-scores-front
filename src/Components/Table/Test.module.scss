.testTable{
    background-color: green;
    color: blue;
}

.testRowClass{
    background-color: yellow;
    color: red;
}