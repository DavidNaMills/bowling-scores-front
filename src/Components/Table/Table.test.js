import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Table from './Table';
import classes from './Test.module.scss';


const tHeaders = ['name', 'Average', 'Pinfall', 'Rank'];

const baseData = [
    {
        id: '',
        values: ['david', 135, 2445, '+ 2'],
        style: {}
    },
    {
        id: '5vd5f3d5cvx3cv5xz53xcz',
        values: ['alan', 189, 2222, '-2'],
        style: {}
    }
]

const tData = [
    {
        id: '',
        values: ['david', 135, 2445, '+ 2'],
        style: { trow: classes.testRowClass }
    },
    {
        id: '5vd5f3d5cvx3cv5xz53xcz',
        values: ['alan', 189, 2222, '-2'],
        style: { trow: { color: 'red' } }
    }
];


const testData = {
    headers: tHeaders,
    rows: tData
}

const testCaption = 'Test label';


describe('<Table /> test suite', () => {

    describe('Snapshot test suite', () => {
        test('matches snapshot with default values', () => {
            const component = renderer.create(<Table data={{
                headers: tHeaders,
                rows: baseData
            }} />);
            const app = component.toJSON();
            expect(app).toMatchSnapshot();
        });

        test('matches snapshot with optional features added', ()=>{
            const component = renderer.create(<Table 
                data={testData}
                showRowNum
            />);
            const app = component.toJSON();
            expect(app).toMatchSnapshot();   
        })
    });

    describe('creation test suite', () => {
        const wrapper = shallow(<Table data={{headers: tHeaders, rows: baseData}} />);

        it('renders a table element', () => {
            const tbl = wrapper.find('table');
            expect(tbl.length).toBe(1);
        });

        it('creates 5 headings', () => {
            const hdr = wrapper.find('thead').find('td');
            const tr = wrapper.find('thead');

            expect(wrapper.find('p').length).toBe(0);
            expect(hdr.length).toBe(tHeaders.length);
            expect(hdr.at(0).text()).toEqual(tHeaders[0]);
            expect(hdr.at(1).text()).toEqual(tHeaders[1]);
            expect(hdr.at(2).text()).toEqual(tHeaders[2]);
            expect(hdr.at(3).text()).toEqual(tHeaders[3]);
            expect(tr.length).toBe(1);
        });

        it('creates 2 rows of table data', () => {
            const tbody = wrapper.find('tbody');
            const tr = wrapper.find('tbody').find('tr');
            const td = wrapper.find('tbody').find('td');

            expect(tbody.length).toBe(1);
            expect(tr.length).toBe(2);
            expect(td.length).toBe(8);
            expect(td.at(0).text()).toEqual('david');
            expect(td.at(4).text()).toEqual('alan');
        });

    });

    describe('styling ', () => {
        //table
        it('adds the default class to the table', () => {
            const wrapper = shallow(<Table data={testData} />);
            expect(wrapper.find('.table').length).toBe(1);
        });

        it('adds a custom class <as String> to the table', () => {
            const wrapper = shallow(<Table data={testData} tableClass={classes.testTable} />);
            expect(wrapper.find(`.${classes.testTable}`).length).toBe(1);
        });

        it('adds a custom className <string> to a specific row', () => {
            const wrapper = shallow(<Table data={testData} />);
            const tr = wrapper.find('tr');
            expect(tr.at(1).find(`.${classes.testRowClass}`).length).toBe(1);

        });

        it('adds a custom inline style to a specific row', () => {
            const wrapper = shallow(<Table data={testData} />);
            const tr = wrapper.find('tr');
            expect(tr.at(2).prop('style')).toHaveProperty('color', 'red');
        });

    });

    describe('options test suite', () => {
        it('contains a <p> tag', ()=>{
            const wrapper = shallow(<Table caption={testCaption} data={testData} />);
            expect(wrapper.find('p').length).toBe(1);
            expect(wrapper.find('p').at(0).text()).toEqual(testCaption);

        });

        it('should display the row number starting at 0', () => {
            const wrapper = shallow(<Table data={testData} showRowNum />);
            const td = wrapper.find('tbody').find('td');

            expect(td.length).toBe(10);
            expect(td.at(0).text()).toEqual('1');
            expect(td.at(5).text()).toEqual('2');
        });

        it('should return the row number of click', () => {
            const mockFn = jest.fn();
            mockFn.mockReturnValue(testData.rows[0].id ? testData.rows[0].id : 1);
            const wrapper = shallow(<Table data={testData} selectRow={mockFn} />);
            wrapper.find('tbody').find('tr').at(0).simulate('click');
            expect(mockFn).toHaveBeenCalled();
            expect(mockFn.mock.results[0].value).toBe(1);
        });

        it('should return the ID supplied when row is clicked', () => {
            const mockFn = jest.fn();
            const wrapper = shallow(<Table data={testData} selectRow={mockFn} />);
            mockFn.mockReturnValue(testData.rows[1].id ? testData.rows[1].id : 1);
            wrapper.find('tbody').find('tr').at(1).simulate('click');
            expect(mockFn).toHaveBeenCalled();
            expect(mockFn.mock.results[0].value).toBe(testData.rows[1].id);
        });
    });

})