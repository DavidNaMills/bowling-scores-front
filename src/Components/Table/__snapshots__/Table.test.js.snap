// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> test suite Snapshot test suite matches snapshot with default values 1`] = `
<div>
  
  <table
    className="table"
  >
    <thead>
      <tr>
        <td>
          name
        </td>
        <td>
          Average
        </td>
        <td>
          Pinfall
        </td>
        <td>
          Rank
        </td>
      </tr>
    </thead>
    <tbody>
      <tr
        className=""
        onClick={[Function]}
        style={Object {}}
      >
        <td>
          david
        </td>
        <td>
          135
        </td>
        <td>
          2445
        </td>
        <td>
          + 2
        </td>
      </tr>
      <tr
        className=""
        onClick={[Function]}
        style={Object {}}
      >
        <td>
          alan
        </td>
        <td>
          189
        </td>
        <td>
          2222
        </td>
        <td>
          -2
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`<Table /> test suite Snapshot test suite matches snapshot with optional features added 1`] = `
<div>
  
  <table
    className="table"
  >
    <thead>
      <tr>
        <td />
        <td>
          name
        </td>
        <td>
          Average
        </td>
        <td>
          Pinfall
        </td>
        <td>
          Rank
        </td>
      </tr>
    </thead>
    <tbody>
      <tr
        className="testRowClass"
        onClick={[Function]}
        style={Object {}}
      >
        <td>
          1
        </td>
        <td>
          david
        </td>
        <td>
          135
        </td>
        <td>
          2445
        </td>
        <td>
          + 2
        </td>
      </tr>
      <tr
        className=""
        onClick={[Function]}
        style={
          Object {
            "color": "red",
          }
        }
      >
        <td>
          2
        </td>
        <td>
          alan
        </td>
        <td>
          189
        </td>
        <td>
          2222
        </td>
        <td>
          -2
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
